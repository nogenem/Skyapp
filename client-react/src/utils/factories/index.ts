import * as models from './models';
import * as states from './states';

export default {
  models,
  states,
};
