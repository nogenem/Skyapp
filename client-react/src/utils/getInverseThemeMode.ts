import { TThemeMode } from '~/redux/theme/types';

// PS: This should be done automatically by the lib... ;/
export default (mode: TThemeMode) => (mode === 'dark' ? 'light' : 'dark');
