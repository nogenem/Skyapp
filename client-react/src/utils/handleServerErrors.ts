interface IErrors {
  [x: string]: string;
}

interface IServerErrorsObject {
  [x: string]: string | string[];
}

interface IServerError {
  errors: IServerErrorsObject;
}

interface IPartialAxiosError {
  response?: {
    data: IServerError;
  };
}

// TODO: Think a way to translate the errors coming from the server
export default (err: IPartialAxiosError): IErrors => {
  if (
    !err ||
    !err.response ||
    !err.response.data ||
    !err.response.data.errors
  ) {
    if (process.env.NODE_ENV !== 'test') console.error(err);
    return { global: 'Internal server error' };
  }

  const { errors } = err.response.data;
  const ret = {} as IErrors;

  Object.keys(errors).forEach((key: string) => {
    const error: string = (
      Array.isArray(errors[key]) ? errors[key][0] : errors[key]
    ) as string;
    ret[key] = error;
  });
  return ret;
};

export type { IErrors, IServerErrorsObject, IServerError, IPartialAxiosError };
