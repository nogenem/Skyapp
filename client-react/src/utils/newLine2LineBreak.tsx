import React from 'react';

// https://stackoverflow.com/questions/40418024/how-to-replace-n-to-linebreaks-in-react-js?rq=1
export default (string: string) =>
  string
    .split('\n')
    .map((item, index) => (index === 0 ? item : [<br key={index} />, item]));
