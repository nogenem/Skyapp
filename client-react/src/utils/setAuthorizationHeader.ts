import { axiosInstance } from '~/services/ApiService';

export default (token?: string) => {
  if (token) {
    axiosInstance.defaults.headers.common.authorization = `Bearer ${token}`;
  } else {
    delete axiosInstance.defaults.headers.common.authorization;
  }
};
