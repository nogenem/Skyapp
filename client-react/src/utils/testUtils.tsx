import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import type { IAppState } from '~/redux/store';

export const getMockStore = () => {
  const middlewares = [thunk];
  return configureStore(middlewares);
};

export const getRenderWithRedux = () => {
  const emptyState: Partial<IAppState> = {};
  const mockStore = getMockStore();

  return (
    ui: React.ReactNode,
    initialState: Partial<IAppState> = emptyState,
  ) => {
    const store = mockStore(initialState);
    return {
      ...render(<Provider store={store}>{ui}</Provider>),
    };
  };
};
