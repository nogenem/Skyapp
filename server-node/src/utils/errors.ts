import {
  INTERNAL_SERVER_ERROR,
  INVALID_ID,
  ROUTE_NOT_FOUND_ERROR,
  INVALID_CREDENTIALS,
  INVALID_EXPIRED_TOKEN,
  NO_USER_WITH_SUCH_EMAIL,
  LAST_EMAIL_SENT_IS_STILL_VALID,
  NO_TOKEN,
  CHANNEL_ALREADY_EXISTS,
  USER_IS_NOT_GROUP_ADM,
  NOT_MEMBER_OF_GROUP,
  CANT_EDIT_THIS_MESSAGE,
} from '~/constants/error_messages';
import type { TTranslatableError } from '~/constants/error_messages';

interface IMsgObj<T> {
  [x: string]: T;
}
type TMsgObjTypes =
  | (string | TTranslatableError)[]
  | string
  | TTranslatableError;

export type { IMsgObj, TMsgObjTypes };

export class CustomError extends Error {
  msgObj: IMsgObj<TMsgObjTypes>;

  kind: string;

  status: number;

  constructor(msgObj: IMsgObj<TMsgObjTypes>, status = 400) {
    super(JSON.stringify(msgObj, null, 2));

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, CustomError);
    }

    this.msgObj = msgObj;
    this.kind = 'CustomError';
    this.status = status;
  }
}

export const validatorErrors = (errs: IMsgObj<TMsgObjTypes>): CustomError =>
  new CustomError(errs);

export const internalServerError = (): CustomError =>
  new CustomError({ global: INTERNAL_SERVER_ERROR }, 500);

export const invalidIdError = (): CustomError =>
  new CustomError({ global: INVALID_ID });

export const mongooseErrors = (errs: IMsgObj<string>): CustomError =>
  new CustomError(errs);

export const routeNotFoundError = (): CustomError =>
  new CustomError({ global: ROUTE_NOT_FOUND_ERROR }, 404);

export const invalidCredentialsError = (): CustomError =>
  new CustomError({ global: INVALID_CREDENTIALS });

export const invalidOrExpiredTokenError = (): CustomError =>
  new CustomError({ global: INVALID_EXPIRED_TOKEN });

export const noUserWithSuchEmailError = (): CustomError =>
  new CustomError({ email: NO_USER_WITH_SUCH_EMAIL });

export const lastEmailSentIsStillValidError = (): CustomError =>
  new CustomError({ email: LAST_EMAIL_SENT_IS_STILL_VALID });

export const noTokenError = (): CustomError =>
  new CustomError({ global: NO_TOKEN });

export const channelAlreadyExistsError = (): CustomError =>
  new CustomError({ global: CHANNEL_ALREADY_EXISTS });

export const userIsNotGroupAdmError = (): CustomError =>
  new CustomError({ global: USER_IS_NOT_GROUP_ADM });

export const notMemberOfGroupError = (): CustomError =>
  new CustomError({ global: NOT_MEMBER_OF_GROUP });

export const multerErrors = (
  errs: [msgObj: IMsgObj<TMsgObjTypes>, status?: number],
): CustomError => new CustomError(...errs);

export const cantEditThisMessageError = (): CustomError =>
  new CustomError({ global: CANT_EDIT_THIS_MESSAGE });
