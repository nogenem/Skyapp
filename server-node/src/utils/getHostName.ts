interface IHeader {
  [x: string]: string | string[] | undefined;
}

export default (headers: IHeader): string => {
  if (headers['x-forwarded-host'])
    return `${headers['x-forwarded-proto']}://${headers['x-forwarded-host']}`;

  if (headers.referer) {
    const host = headers.referer as string;
    const url = new URL(host);
    return url.origin;
  }

  let { host } = headers;
  host = (host || '') as string;
  if (!host.startsWith('http')) host = `http://${host}`;
  return host;
};
