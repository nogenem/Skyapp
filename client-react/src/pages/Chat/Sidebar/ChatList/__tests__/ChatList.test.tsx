import React from 'react';

import type { IChannel } from '~/redux/chat/types';
import FACTORIES from '~/utils/factories';
import { getRenderWithRedux } from '~/utils/testUtils';

import { ChatList } from '../index';

const renderWithRedux = getRenderWithRedux();

describe('Connected ChatList', () => {
  it('renders correctly', () => {
    const initialState = {
      chat: FACTORIES.states.chat({}, { useConstValues: true }),
    };

    const { container } = renderWithRedux(
      <ChatList filter="" activeChannelId={undefined} />,
      initialState,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders correctly when `filter` is not empty', () => {
    const channel1: IChannel = FACTORIES.models.channel(
      {
        _id: 'channel-1',
        name: 'Channel 1',
      },
      { useConstValues: true },
    );
    const channel2: IChannel = FACTORIES.models.channel(
      {
        _id: 'channel-2',
        name: 'Channel 2',
      },
      { useConstValues: true },
    );
    const initialState = {
      chat: FACTORIES.states.chat({
        channels: {
          [channel1._id]: channel1,
          [channel2._id]: channel2,
        },
      }),
    };

    const { queryByText } = renderWithRedux(
      <ChatList filter={channel1.name} activeChannelId={undefined} />,
      initialState,
    );

    const channel1NameRegExp = new RegExp(channel1.name, 'i');
    const channel2NameRegExp = new RegExp(channel2.name, 'i');
    expect(queryByText(channel1NameRegExp)).toBeInTheDocument();
    expect(queryByText(channel2NameRegExp)).not.toBeInTheDocument();
  });
});
