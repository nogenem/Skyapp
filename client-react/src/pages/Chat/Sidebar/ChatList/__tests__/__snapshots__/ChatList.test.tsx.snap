// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected ChatList renders correctly 1`] = `
<div>
  <h6
    class="MuiTypography-root makeStyles-chatsText-1 MuiTypography-subtitle2"
  >
    Common:CHATS
  </h6>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="makeStyles-avatarContainer-2"
      >
        <div
          class="makeStyles-container-7"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-8"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </div>
          <span
            class="makeStyles-dotContainer-9"
          >
            <span
              class="makeStyles-dot-10 makeStyles-dot_active-11 "
              title="Common:Active"
            />
          </span>
        </div>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          <span
            class="makeStyles-textContainer-3"
            title="Channel 1"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-noWrap MuiTypography-displayInline"
            >
              Channel 1
            </span>
            <span
              class="MuiTypography-root MuiTypography-caption MuiTypography-colorInherit MuiTypography-displayInline"
            >
              04/12/2021
            </span>
          </span>
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          <span
            class="makeStyles-textContainer-3"
            title="Test message 1"
          >
            <span
              class="MuiTypography-root makeStyles-secondaryText-4 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-noWrap MuiTypography-displayInline"
            >
              Test message 1
            </span>
            <span
              class="MuiTypography-root MuiTypography-caption MuiTypography-colorInherit MuiTypography-displayInline"
            >
              <span
                class="MuiBadge-root makeStyles-unreadBadge-6"
              >
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
                >
                  2
                </span>
              </span>
            </span>
          </span>
        </p>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
    <div
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
      role="button"
      tabindex="0"
    >
      <div
        class="makeStyles-avatarContainer-2"
      >
        <div
          class="makeStyles-container-7"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-8"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </div>
          <span
            class="makeStyles-dotContainer-9"
          >
            <span
              class="makeStyles-dot-10 makeStyles-dot_active-11 "
              title="Common:Active"
            />
          </span>
        </div>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          <span
            class="makeStyles-textContainer-3"
            title="Channel 1"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-noWrap MuiTypography-displayInline"
            >
              Channel 1
            </span>
            <span
              class="MuiTypography-root MuiTypography-caption MuiTypography-colorInherit MuiTypography-displayInline"
            >
              04/12/2021
            </span>
          </span>
        </span>
        <p
          class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          <span
            class="makeStyles-textContainer-3"
            title="Test message 1"
          >
            <span
              class="MuiTypography-root makeStyles-secondaryText-4 MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-noWrap MuiTypography-displayInline"
            >
              Test message 1
            </span>
            <span
              class="MuiTypography-root MuiTypography-caption MuiTypography-colorInherit MuiTypography-displayInline"
            >
              <span
                class="MuiBadge-root makeStyles-unreadBadge-6"
              >
                <span
                  class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
                >
                  2
                </span>
              </span>
            </span>
          </span>
        </p>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </div>
  </ul>
</div>
`;
